### PR TITLE
#882 dragging keyframe displacement

### LINKED ISSUE #882: [UX][Bug] Dragging keyframe along timeline should not move other keyframes if not dropped on top
### --Issue Summary--
This is both a request for a fix and a discussion for enhancement. I noticed this problem a long time ago, but I forgot to report it. When moving frame containers around the timeline will displace other keyframes from their original intended position and adds 1 more frame to the opposite side of the direction you were dragging the initial drawing from.

### --Expected Results--
When moving a keyframe over another, any keyframe that is not being dragged should not move from their current position unless you drop it on top of that keyframe.

**Discussion:** Please not that this shifting behavior however is USEFUL when you are moving more than one keyframe at the same time, because you will displace keyframes to make room so it shouldn't go away, but the user should be able to control this. 

So if possible it could work on a toggle via shortcut, similar to how TVPaint does it. So we could have the move and slide (current behaviour but fixing the additional 1 frame skip) and the move and replace which is basically replacing the keyframe below with which ever is dropped directly on top (this taking into account the possibility of dragging multiple drawing containers on top, so the one that matches with the container already there would be replaced)

**For REFerence** In TVPaint you can Move (and overwrite), Displace (Insert) and Copy an image to any frame. Please read this to understand the concept: https://www.tvpaint.com/doc/tvp11/index.php?id=lesson-animate-first-steps-selection-navigation-movement-move-image-inside-layer

So, for example, we can move and not shift keyframes if they are less than 2, but if you have more than two selected and dragged over, you'd have to press CTRL to activate the sliding behaviour.

Of course the current work around is to move the frames that got shifted back, but when you have a long animation you're working on it becomes inefficient.

### --Steps to reproduce--
1) add 2 keyframes. You'll have 3 including the first that's always created with Pencil2D
2) Position the keyframes along the timeline like this (it doesn't matter the position, but this is to create a reference so you can notice the issue):
+ Keyframe 1 -> Frame 1
+ Keyframe 2 -> Frame 4
+ Keyframe 3 -> Frame 8
3) Now take the last keyframe and move it to frame 3. Notice how the keyframe in frame 4, was move to frame 5. It should stay in frame 4 unless I explicitly dropped the keyframe I was dragging over frame 4.
4) Take the keyframe on frame 3 and take it over again to frame 8. Notice how the keyframe in frame 5 will go back to frame 4.
5) Just for completeness sake take the key on frame 1 and move it to frame 5. Notice the one on frame 4 will go to frame 3. (Even if it's a duplicate, the keyframes are shifting unreasonably)

### --System Information--
+ Pencil2D Version:
Version: 0.6.0 Development build

+ Operating System:
Windows 7

### PR DESCRIPTION
**Pull request by MrStevns**
I've rewritten the frame dragging functionality so frames can never be misplaced by accident, this means that frames will be dragged across other frames and only moved when the mouse has been released and there was space for the frames, otherwise the frames will be put back.

### Changes
- Dragging is cosmetic until the very moment you let go on the frames. 
- Frames are slightly lifted off the track to help visualize that dragging has been initiated.'

### What's removed
- Ability to easily swap frames when dragging, eg. the old functionality. 

Result:
displacement

The intend of dragging should be explicit and in combination with https://github.com/pencil2d/pencil/pull/1343, dragging and manipulating frames in general will become more intuitive.

closes #882

### CODE REVIEW

**Comment by CandyFace** [comment]
Code is up to date and can now be reviewed.

File: core_lib/src/interface/timelinecells.cpp [hunk 1]
```diff
@@ -857,14 +903,29 @@ void TimeLineCells::mouseReleaseEvent(QMouseEvent* event)
...
-    if (frameNumber < 1) frameNumber = -1;
+    if (frameNumber < 1) frameNumber = 1;
     int layerNumber = getLayerNumber(event->pos().y());
-    if (mType == TIMELINE_CELL_TYPE::Tracks && primaryButton != Qt::MidButton && layerNumber != -1 && layerNumber < mEditor->object()->getLayerCount())
+    if (mType == TIMELINE_CELL_TYPE::Tracks && primaryButton != Qt::MidButton)
     {
-        Layer *currentLayer = mEditor->object()->getLayer(layerNumber);
+        // We should affect the current layer based on what's selected, not where the mouse currently is.
+        Layer *currentLayer = mEditor->layers()->getLayer(mCurrentLayerNumber);
+        if (currentLayer == nullptr) { return; }
```
Comment (J5lx) [inline core_lib/src/interface/timelinecells.cpp:914]: This return statement prevents later code in this method from running. Is that intentional?

**Review by J5lx — CHANGES_REQUESTED** [review]
Looks pretty good overall! There’s just one bug where the offset of the frames being dragged gets messed up when the contents of the timeline are moved to the left by scrolling. Other than that there are just a handful of remarks and the usual PR for the cosmetic stuff (CandyFace#18). Nice work!

File: core_lib/src/structure/layer.cpp [hunk 2]
```diff
@@ -435,86 +434,75 @@ void Layer::deselectAll()
...
-        // Update selection lists
-        for (int i = 0; i < mSelectedFrames_byPosition.count(); i++)
-        {
-            mSelectedFrames_byPosition[i] = mSelectedFrames_byPosition[i] + offset;
-        }
-        for (int i = 0; i < mSelectedFrames_byLast.count(); i++)
+        // Get the frame to move
+        KeyFrame* selectedFrame = getKeyFrameAt(fromPos);
+
+        if (selectedFrame != nullptr && !keyExists(toPos))
```
[the remark this answers is not in the document]
Comment (CandyFace) [reply core_lib/src/structure/layer.cpp:482]: Good point, yes it should. Will change.

File: core_lib/src/interface/timelinecells.cpp [hunk 3]
```diff
@@ -201,47 +203,52 @@ void TimeLineCells::drawContent()
     painter.setBrush(palette.color(QPalette::Base));
     painter.drawRect(QRect(0, 0, width(), height()));
 
+    const int widgetWidth = width();
+    const int layerHeight = mLayerHeight;
```
[the remark this answers is not in the document]
Comment (CandyFace) [reply core_lib/src/interface/timelinecells.cpp:204]: it depends, from a pure readability perspective, it doesn't have any advantage, however if I were to rename or change the value.. then i would have to do it 6 places.

At the time, while refactoring and working on the hover logic I found it easier to keep the local variable when tweaking things but I have no strong feelings towards it.

File: core_lib/src/interface/timelinecells.cpp (hunk 1 above)
Comment (CandyFace) [reply core_lib/src/interface/timelinecells.cpp:910]: Hmm most likely not.. that seems to be an afterthought. Will change.

File: core_lib/src/interface/timelinecells.cpp [hunk 4]
```diff
@@ -857,14 +903,29 @@ void TimeLineCells::mouseReleaseEvent(QMouseEvent* event)
...
-        Layer *currentLayer = mEditor->object()->getLayer(layerNumber);
+        // We should affect the current layer based on what's selected, not where the mouse currently is.
+        Layer *currentLayer = mEditor->layers()->getLayer(mCurrentLayerNumber);
+        if (currentLayer == nullptr) { return; }
 
-        if (!mTimeLine->scrubbing && !mMovingFrames && !mClickSelecting && !mBoxSelecting)
+        if (mMovingFrames) {
+
+            // New offset is first added after 50% of the frame has been crossed
+            int frameNumber = getFrameNumber(event->pos().x());
```
[the remark this answers is not in the document]
Comment (CandyFace) [reply core_lib/src/interface/timelinecells.cpp:915]: It's an oversight.. will remove.

**Comment by CandyFace** [comment]
Ah nice spotted, I indeed hadn't accounted for the scroll offset when painting the moving frames! 
PR remarks handled and branch is now also up to date with master.

File: core_lib/src/interface/timelinecells.cpp [hunk 5]
```diff
@@ -23,6 +23,9 @@ GNU General Public License for more details.
 #include <QPainter>
 #include <QSettings>
 #include <QMenu>
+
+#include <QDebug>
```
Comment (J5lx) [inline core_lib/src/interface/timelinecells.cpp:27]: leftover include

**Review by J5lx — CHANGES_REQUESTED** [review]
I think it would make sense to hide the hover “ghost frame” while dragging keyframes, which wasn’t an issue before since #1343 hadn’t been merged yet. Other than that this PR is now pretty much good to go I’d say. Nice job!

**Comment by MrStevns** [comment]
PR updated and remarks handled. Ready for takeoff!

**Review by J5lx — APPROVED** [review]
Awesome! Let’s go!

**Pull request merged**